awscli/autoprompt/core.py

```diff
@@ -29,47 +32,54 @@ class AutoPromptDriver:
 
     NO_PROMPT_ARGS = ['help', '--version']
 
-    def __init__(self, session, prompter=None):
+    def __init__(self, driver, completion_source=None, prompter=None):
+        self._completion_source = completion_source
         self._prompter = prompter
-        self._session = session
-
-    def _should_autoprompt(self, parsed_args, args):
+        self._session = driver.session
+        self._driver = driver
+        if self._completion_source is None:
+            self._completion_source = create_autocompleter(driver=self._driver)
+
+    @property
+    def prompter(self):
+        if self._prompter is None:
+            self._prompter = AutoPrompter(self._completion_source,
+                                          self._driver)
+        return self._prompter
+
+    def _filter_out_options(self, args):
+        override_options = ['--cli-auto-prompt', '--no-cli-auto-prompt']
+        return [
+                arg for arg in args
+                if (not arg.startswith('--') or arg in override_options)
+            ]
+
+    def resolve_mode(self, args):
         # Order of precedence to check:
         # - check if any arg rom NO_PROMPT_ARGS in args
         # - check if '--no-cli-auto-prompt' was specified
         # - check if '--cli-auto-prompt' was specified
         # - check configuration chain
+        option_parser = ArgTableArgParser(self._driver.arg_table)
+        filtered_args = self._filter_out_options(args)
+        parsed_options, _ = option_parser.parse_known_args(filtered_args)
+        validate_auto_prompt_args_are_mutually_exclusive(parsed_options)
         if any(arg in args for arg in self.NO_PROMPT_ARGS):
-            return False
-        if getattr(parsed_args, 'no_cli_auto_prompt', False):
-            return False
-        if getattr(parsed_args, 'cli_auto_prompt', False):
-            return True
+            return 'off'
+        if getattr(parsed_options, 'no_cli_auto_prompt', False):
+            return 'off'
+        if getattr(parsed_options, 'cli_auto_prompt', False):
+            return 'on'
         config = self._session.get_config_variable('cli_auto_prompt')
```

Comment:
One bug that I think we will need to address is when we resolve `cli_auto_prompt` we do not take into account `--profile`. For example, if I'm using my `default` profile and and specify a `--profile other-profile` that does not have auto-prompt configured, I will still get auto-prompt mode. This may be a little tricky to fix and may need to make a larger change. So I'm happy with following up on this in another pull request if it requires larger changes